process/transaction/shardProcess.go

```diff
@@ -235,15 +235,28 @@ func (txProc *txProcessor) createReceiptWithReturnedGas(tx *transaction.Transact
 	return nil
 }
 
-func (txProc *txProcessor) processTxFee(tx *transaction.Transaction, acntSnd state.UserAccountHandler) (*big.Int, error) {
-	if acntSnd == nil {
+func (txProc *txProcessor) processTxFee(
+	tx *transaction.Transaction,
+	acntSnd, acntDst state.UserAccountHandler,
+) (*big.Int, error) {
+	if check.IfNil(acntSnd) {
 		return big.NewInt(0), nil
 	}
 
+	totalCost := big.NewInt(0).Mul(big.NewInt(0).SetUint64(tx.GetGasLimit()), big.NewInt(0).SetUint64(tx.GetGasPrice()))
 	cost := txProc.economicsFee.ComputeFee(tx)
-	err := acntSnd.AddToBalance(big.NewInt(0).Neg(cost))
-	if err != nil {
-		return nil, err
+
+	isCrossShardSCCall := check.IfNil(acntDst) && len(tx.GetData()) > 0 && core.IsSmartContractAddress(tx.GetRecvAddress())
+	if isCrossShardSCCall {
+		err := acntSnd.AddToBalance(big.NewInt(0).Neg(totalCost))
+		if err != nil {
+			return nil, err
+		}
+	} else {
+		err := acntSnd.AddToBalance(big.NewInt(0).Neg(cost))
```

Comment:
the actual cost for the current shard is only the moveBalance. the rest goes to the destination shard and it is added there to the accumulated fees. In case of simple move balance cross shard, the only cost the sender shard has to take out is the gasPrice * moveBalanceGasCost. In case of smart contract, all the gasCost has to be taken out, as you will get back the unused part as a smart contract result. In case of cross-shard movebalance transactions there is no receipt coming back with the unused gas - as you can precalculate.

Reply:
ok